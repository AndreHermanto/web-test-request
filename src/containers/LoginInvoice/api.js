/**
 * Returns an invoice. 
 */
export function getInvoice(user) {
  return fetch(`${process.env.REACT_APP_BASE_URL}/test-requests?id=${user.id}&mobilePhone=${user.phone}&lastName=${user.lastName}`)
    .then((response) => response.json())
    .then((json) => json.data);
}
