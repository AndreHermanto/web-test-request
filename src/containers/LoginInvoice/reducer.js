import validator from './../../components/validator';

export function initData(){
  var state = {
    form:{
      phone: "",
      lastName: ""
    },
    validationRule: {
      phone: 'required number',
      lastName: 'required'
    },
    validated: false,
    authenticated: true
  }
    
    // This validates the data in the initial state.
    state.validation = {};
    for (var field in state.validationRule) {
      if(field) {
        state.validation[field] = validator(state.form[field], state.validationRule[field]);
      }
    }
    
    return Object.assign({}, state);
}

/**
 * setFormState
 */
export function setFormState(state, target) {
  var value;
  switch(target.type) {
    case 'checkbox':
      value = target.checked;
      break;
    default:
      value = target.value;
      break;
  } 
  
  var formStateChild = Object.assign({}, state.form, {
    [target.name]: value
  });
  
  var validationStateChild = Object.assign({}, state.validation, {
    [target.name]: validator(value, state.validationRule[target.name])
  });

  return Object.assign({}, state, {
    form: formStateChild,
    validation: validationStateChild
  });
}

/**
 * Set "validated" state to true - identifying the confirm button is clicked and validation processed.
 * @param {Object} state Targeted state to be changed.
 */
export function validatedToTrue(state) {
  return Object.assign({}, state, {
    validated: true
  });
}

/**
 * Set "authenticated" state to false - identifying the confirm button is clicked and user is valid.
 * @param {Object} state Targeted state to be changed.
 */
export function authenticatedToFalse(state) {
  return Object.assign({}, state, {
    authenticated: false
  });
}