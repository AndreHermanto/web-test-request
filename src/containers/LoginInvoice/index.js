import React, { Component } from 'react';
import { 
  Row,
  Col,
  Form,
  Glyphicon
} from 'react-bootstrap';
import styled from 'styled-components';
import { 
  initData,
  setFormState,
  validatedToTrue,
  authenticatedToFalse } from './reducer';
import { getInvoice } from './api';
import { hashHistory } from 'react-router';
import InputInvoice from './../../components/InputInvoice';

const FormContainer = styled.div`
  height: 90px; 
`;

const Container = styled.div`
  margin-top: 50px;
`;

const SubmitButton = styled.button`
  border-style: solid;
  border-width: 0.01px;
  background-color: #00a6b6;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 16px;
  border-radius: 0px;
  -webkit-transition: background-color 0.5s;
  -webkit-transition-timing-function: ease;
  text-transform: uppercase;
  &:hover {
    background-color: #00c9dc;
  }
`;

const Login = styled.div`
  width: 100%;
  height: 400px;
  border-style: solid;
  padding: 40px 40px 40px 50px;
  text-align: center;
  background-color: #f6f6f6;
  border-width: 3px;
  border-color: #00525a;
  color: #00525a;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
`;
const Invoice = styled.h3`
  width:130px;
  margin-top:-57px;
  margin-left: 31%;
  margin-bottom: 30px;
  background:white;
`;

const Warning = styled.div`
  height: 15px;
  color: #a94442;
  margin-bottom: 20px;
`;

/**
 * LoginInvoice - Landing page to see invoice
 */
class LoginInvoice extends Component {
  constructor(props) {
    super(props);
    this.state = initData();

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleConfirm = this.handleConfirm.bind(this);

	}
  handleInputChange(event) {
    const target = event.target;
    this.setState(setFormState(this.state, target));
  }

  handleSubmit(passValidation) {
    if(!passValidation) return false;
    var self = this;
    getInvoice(Object.assign({}, this.state.form, {id: this.props.params.id}))
    .then(function(response) {
      if(response.length){
        hashHistory.push({
          pathname: '/invoice',
          query: {data: JSON.stringify(response[0]).data}
        })
      }else{
        self.setState(authenticatedToFalse(self.state));
      }
    })
  }

  handleConfirm() {
    return this.setState(validatedToTrue(this.state), () => {    
      var pass = true;
      for (var field in this.state.validation) {
        if(!this.state.validation[field].skip) {
          if(this.state.validation[field].status === 'error') pass = false;
        }
      }

      this.handleSubmit(pass); 
    });
  }
  // This renders the validation result after confirm button is clicked.
  validate() {
    return this.state.validated && this.state.validation;
  }

	render(){
    return(
    <Container>
      <Row>
        <Col xs={6} xsOffset={3}>
          <Login>     
            <Invoice><Glyphicon glyph="glyphicon glyphicon-list-alt" /> INVOICE</Invoice>
            <Form>
              <FormContainer>              
                <InputInvoice
                  field="lastName"
                  label="Last Name"
                  onChange={this.handleInputChange}
                  onValidate={this.validate()}
                  formState={this.state.form}
                  glyphicon="glyphicon-user"
                />
              </FormContainer>
              <FormContainer>
                <InputInvoice
                  field="phone"
                  label="Phone"
                  onChange={this.handleInputChange}
                  onValidate={this.validate()}
                  formState={this.state.form}
                  glyphicon="glyphicon-phone-alt"
                />
              </FormContainer>
              <Warning>{this.state.authenticated? null : "Invalid phone number or last name"}</Warning>
              <SubmitButton type="button" onClick={this.handleConfirm}>
                View Invoice
              </SubmitButton>
            </Form>
          </Login>
        </Col>
      </Row>
    </Container>
	  )
  }
}

export default LoginInvoice;