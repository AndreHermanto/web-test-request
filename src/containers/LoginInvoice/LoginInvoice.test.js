import React from 'react';
import ReactDOM from 'react-dom';
import LoginInvoice from './index';
import renderer from 'react-test-renderer';

describe('Create new ontology page', () => {
  it('LoginInvoice', () => {
    const div = document.createElement('div');
    ReactDOM.render(<LoginInvoice />, div);
  });

  it('LoginInvoice renders correctly', function() {
    const tree = renderer.create(<LoginInvoice />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});