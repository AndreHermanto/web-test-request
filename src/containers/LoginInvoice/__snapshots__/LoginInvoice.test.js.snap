exports[`Create new ontology page LoginInvoice renders correctly 1`] = `
<div
  className="gZxEPq">
  <div
    className="row">
    <div
      className="col-xs-6 col-xs-offset-3">
      <div
        className="blkGcf">
        <h3
          className="jQRRzI">
          <span
            className="glyphicon glyphicon-glyphicon glyphicon-list-alt" />
           INVOICE
        </h3>
        <form
          className="">
          <div
            className="bfUfcU">
            <div
              className="form-group has-feedback">
              <label
                className="control-label"
                htmlFor={undefined}>
                Last Name
              </label>
              <input
                className="form-control"
                disabled={undefined}
                id={undefined}
                name="lastName"
                onChange={[Function]}
                placeholder="Your Last Name"
                type="text"
                value="" />
              <span
                className="form-control-feedback epmpPI glyphicon glyphicon-glyphicon glyphicon-user" />
              <span
                className="SfhHK help-block" />
            </div>
          </div>
          <div
            className="bfUfcU">
            <div
              className="form-group has-feedback">
              <label
                className="control-label"
                htmlFor={undefined}>
                Phone
              </label>
              <input
                className="form-control"
                disabled={undefined}
                id={undefined}
                name="phone"
                onChange={[Function]}
                placeholder="Your Phone"
                type="text"
                value="" />
              <span
                className="form-control-feedback epmpPI glyphicon glyphicon-glyphicon glyphicon-phone-alt" />
              <span
                className="SfhHK help-block" />
            </div>
          </div>
          <div
            className="jhqAqZ" />
          <button
            className="btTVNa"
            onClick={[Function]}
            type="button">
            View Invoice
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
